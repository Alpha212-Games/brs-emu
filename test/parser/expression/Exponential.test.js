const Parser = require("../../../lib/parser");
const { Lexeme } = require("../../../lib/Lexeme");
const BrsError = require("../../../lib/Error");

const { token, EOF } = require("../ParserTests");

describe("parser", () => {
    afterEach(() => BrsError.reset());

    describe("exponential expressions", () => {
        it("parses exponential operators", () => {
            let parsed = Parser.parse([
                token(Lexeme.Integer, 2),
                token(Lexeme.Caret),
                token(Lexeme.Integer, 3),
                EOF
            ]);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses repeated exponential operators as left-associative", () => {
            let parsed = Parser.parse([
                token(Lexeme.Integer, 2),
                token(Lexeme.Caret),
                token(Lexeme.Integer, 3),
                token(Lexeme.Caret),
                token(Lexeme.Integer, 4),
                EOF
            ]);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });
    });
});