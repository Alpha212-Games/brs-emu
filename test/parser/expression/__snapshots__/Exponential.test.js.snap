// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser exponential expressions parses exponential operators 1`] = `
Array [
  Assignment {
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "_",
    },
    "value": Binary {
      "left": Literal {
        "value": Int32 {
          "kind": 3,
          "value": 2,
        },
      },
      "right": Literal {
        "value": Int32 {
          "kind": 3,
          "value": 3,
        },
      },
      "token": Object {
        "kind": 6,
        "line": 1,
        "text": "^",
      },
    },
  },
]
`;

exports[`parser exponential expressions parses repeated exponential operators as left-associative 1`] = `
Array [
  Assignment {
    "name": Object {
      "kind": 21,
      "line": 1,
      "text": "_",
    },
    "value": Binary {
      "left": Binary {
        "left": Literal {
          "value": Int32 {
            "kind": 3,
            "value": 2,
          },
        },
        "right": Literal {
          "value": Int32 {
            "kind": 3,
            "value": 3,
          },
        },
        "token": Object {
          "kind": 6,
          "line": 1,
          "text": "^",
        },
      },
      "right": Literal {
        "value": Int32 {
          "kind": 3,
          "value": 4,
        },
      },
      "token": Object {
        "kind": 6,
        "line": 1,
        "text": "^",
      },
    },
  },
]
`;
